### PR TITLE
fix segmentation fault when changing password with chpasswd

### PR DESCRIPTION
**Pull request by juyin**
Returns null when fopen fails. Then, using fread with a null pointer will cause a segfault.

### CODE REVIEW

**Comment by juyin** [comment]
The issue： #514

**Comment by ikerexxe** [comment]
LGTM but since I proposed the solution I prefer that somebody else takes a look at the solution.

Thanks for the patch!

File: src/chpasswd.c [hunk 1]
```diff
@@ -392,12 +392,49 @@ static void close_files (void)
 	pw_locked = false;
 }
 
+static const char *get_salt(void)
+{
+	if (   !eflg
+		&& (   (NULL == crypt_method)
+			|| (0 != strcmp (crypt_method, "NONE")))) {
```
Comment (ikerexxe) [inline src/chpasswd.c:399]: As you are changing this part of the code. Can you make it easier to read?

File: src/chpasswd.c [hunk 2]
```diff
@@ -392,12 +392,49 @@ static void close_files (void)
...
+		void *arg = NULL;
+
+		if (md5flg) {
+			crypt_method = "MD5";
+		}
+#if defined(USE_SHA_CRYPT) || defined(USE_BCRYPT) || defined(USE_YESCRYPT)
+		if (sflg) {
+#if defined(USE_SHA_CRYPT)
+			if (   (0 == strcmp (crypt_method, "SHA256"))
+				|| (0 == strcmp (crypt_method, "SHA512"))) {
```
Comment (ikerexxe) [inline src/chpasswd.c:409]: Make this part also easier to read.

**Review by ikerexxe — CHANGES_REQUESTED** [review]
I have added two minor comments.

File: src/chpasswd.c (hunk 1 above)
Comment (juyin) [reply src/chpasswd.c:399]: I have not changed the code. I just moved this part of code in a function.

Comment (ikerexxe) [reply src/chpasswd.c:399]: This is more about the spaces/tab between `if (` and `!eflg`. I don't know why the original author did that but IMO it's difficult to read. Can you change it to something like:
```
	if (!eflg
	    && ((NULL == crypt_method) || (0 != strcmp (crypt_method, "NONE")))) {
```

This is the best I could come up with, but I am open to better solutions.

Comment (juyin) [reply src/chpasswd.c]: > This is more about the spaces/tab between `if (` and `!eflg`. I don't know why the original author did that but IMO it's difficult to read. Can you change it to something like:
> 
> ```
> 	if (!eflg
> 	    && ((NULL == crypt_method) || (0 != strcmp (crypt_method, "NONE")))) {
> ```
> 
> This is the best I could come up with, but I am open to better solutions.

OK, I have refactored this function

**Review by ikerexxe — APPROVED** [review]
LGTM! Thanks for taking care of it.

**Comment by hallyn** [comment]
It seems to me there are cases of strcmp(crypt_method, X) being called the code before this patch.  Could you fix those in this PR as well?  In either case the last patch is a nice cleanup but challenging to reason about, so I want to take another look after a break.

**Comment by juyin** [comment]
> It seems to me there are cases of strcmp(crypt_method, X) being called the code before this patch. Could you fix those in this PR as well? In either case the last patch is a nice cleanup but challenging to reason about, so I want to take another look after a break.

OK, we can replace it with a macro to make it cleanup.

**Comment by hallyn** [comment]
That would be great - but could you have the macro first check for str not being NULL? :)

**Comment by hallyn** [comment]
Oh, never mind, cflg handles it.

**Comment by hallyn** [comment]
> Oh, never mind, cflg handles it.

No, it doesn't.  So I figured

```
#define IS_CRYPT_METHOD(str) ((crypt_method != NULL && strcmp(crypt_method, str) == 0) ? true : false)
```

**Comment by hallyn** [comment]
Thanks!

**Pull request merged**